src/grader_webhook.ts

```diff
@@ -0,0 +1,415 @@
+  if (null === task) {
+    throw new Error(`Cannot find task ${submission.idTask}`);
+  }
+
+  const allTests = await Db.execute<TaskTest[]>(`SELECT tm_tasks_tests.sName, tm_tasks_tests.ID, tm_tasks_tests.sGroupType, tm_tasks_tests.iRank, tm_tasks_tests.sOutput from tm_tasks_tests
+    JOIN tm_submissions ON tm_submissions.idTask = tm_tasks_tests.idTask
+    WHERE tm_submissions.ID = :idSubmission and ((tm_tasks_tests.sGroupType = 'Evaluation' or tm_tasks_tests.sGroupType = 'Submission') or (tm_tasks_tests.sGroupType = 'User' and tm_tasks_tests.idUser = tm_submissions.idUser and tm_tasks_tests.idPlatform = tm_submissions.idPlatform));`,
+  {
+    idSubmission: tokenParams.sTaskName
+  });
+
+  const testsByName: {[key: string]: TaskTest} = {};
+  // This is a bit odd: when submitting answer for evaluation, the resulting
+  // json contains the names of the tests, but when the submission is with user tests,
+  // it contains names like id-xxx where xxx is the ID of the tm_task_test
+  for (const test of allTests) {
+    if ('Submission' === test.sGroupType) {
+      testsByName[`id-${test.ID}`] = test;
+    } else {
+      testsByName[test.sName] = test;
+    }
+  }
+
+  const graderResults = tokenParams.sResultData;
+  let nbTestsPassed = 0;
+  let iScoreTotal = 0;
+  let nbTestsTotal = 0;
+  let bCompilError = false;
+  let sCompilMsg = graderResults?.solutions[0]?.compilationExecution?.stderr?.data ? graderResults?.solutions[0]?.compilationExecution?.stderr.data : '';
+  let sErrorMsg = '';
+  let iScore = 0;
+
+  if (task.bTestMode) {
+    sCompilMsg = '';
+    let nbTestsFailedTotal = 0;
+    const invalidTests: {[testName: string]: boolean} = {};
+    for (const execution of graderResults.executions) {
+      nbTestsTotal++;
+      if (!(execution.name in testsByName)) {
+        testsByName[execution.name] = await createNewTest(tokenParams.sTaskName, task.ID, execution.name);
+      }
+
+      const test = testsByName[execution.name];
+      let nbTestFailed = 0;
+      for (const testReport of execution.testsReports) {
+        const testName = testReport.name.substring(5);
+        if (!testReport.checker) {
+          if (!(testName in invalidTests)) {
+            invalidTests[testName] = true;
+            bCompilError = true;
+            const thisCompilMsg = testReport.sanitizer.stdout.data;
+            sCompilMsg += `
+Erreur dans le test : ${testName}
+${thisCompilMsg}`;
+          }
+        } else {
+          const outData = testReport.checker.stdout.data;
+          const iScore = Number(outData.split('\n')[0]);
+          if (0 === iScore) {
+            nbTestFailed++;
+          }
+        }
+      }
+
+      let iErrorCode: number;
+      let sLog: string;
+      if (nbTestFailed) {
+        nbTestsFailedTotal++;
+        iScore = 100;
+        iErrorCode = 0;
+        sLog = `${nbTestFailed} de vos tests permet${nbTestFailed > 1 ? 'tent' : ''} de détecter l'erreur de cette solution.`;
```

Comment:
These projects are designed to be used for a broader usage in the medium-term, so I think (I may be wrong) that messages in French should always be put in English (or i18n of course)...  otherwise, from my experience of using franceioi tools in other countries, we always have a message popping in FR and we don't know why and it is very annoying for non-FR-speakers.

Reply:
My answer is not replying to what you say, but so my point is 1) put them in English, 2) see if it is really useful to translate them (typically in a frontend-backend pattern, backend does not translate anything, i18n is a frontend thing)